### PR TITLE
 Boundary slope control in shape optimization for additive manufacturing: for self-support and surface roughness

### PR DESCRIPTION
**Pull request by RezaNajian**
**📝 Description**
The new responses control boundary slope of optimized parts in node-based shape optimization for additive manufacturing (AM). Boundary slope of a part affects the amount of support structure required during its fabrication by additive processes. Boundary slope also has direct relation with the resulting surface roughness from the AM processes, which in turn affects the heat transfer efficiency. By constraining the minimal boundary slope, support structures can be eliminated or reduced for AM, and thus material and post-processing costs are reduced; by constraining the maximal boundary slope, high surface roughness can be attained, and thus the heat transfer efficiency is increased. The overhang angle constraints are cast to a global/aggregated constraint through a Heaviside projection. a Heaviside Projection based Integral (HPI) formulation is applied. The main idea of this formulation is that the local constraints are projected by a Heaviside function to nonnegative numbers and integrated over the domain, then the integral is constrained to eliminate the violated constraints. Through this formulation, a single global constraint can efficiently control local boundary slope.

Please mark the PR with appropriate tags: 
- ShapeOptimizationApplication

**🆕 Changelog**
Please summarize the changes in one list to generate the changelog:
E.g.
- Added AM_boundary_roughness_angle_response_function and AM_boundary_roughness_angle_response_function
- Added new nodal vars NODAL_MIN_ANGLE, NODAL_MAX_ANGLE
- Added test case

### CODE REVIEW

**Comment by philbucher** [comment]
seems like you accidentally committed some unwanted things like the git folder and the examples repo

File: applications/ShapeOptimizationApplication/custom_python/add_custom_utilities_to_python.cpp [hunk 1]
```diff
@@ -34,6 +34,8 @@
 #include "custom_utilities/input_output/universal_file_io.h"
 #include "custom_utilities/search_based_functions.h"
 #include "custom_utilities/response_functions/face_angle_response_function_utility.h"
+#include "custom_utilities/response_functions/AM_boundary_slope_angle_response_function_utility.h"
+#include "custom_utilities/response_functions/AM_boundary_roughness_angle_response_function_utility.h"
```
Comment (philbucher) [inline applications/ShapeOptimizationApplication/custom_python/add_custom_utilities_to_python.cpp]:
```suggestion
#include "custom_utilities/response_functions/am_boundary_slope_angle_response_function_utility.h"
#include "custom_utilities/response_functions/am_boundary_roughness_angle_response_function_utility.h"
```

lower case is the naming convention

**Comment by RezaNajian** [comment]
> seems like you accidentally committed some unwanted things like the git folder and the examples repo

right. I corrected the issues.

File: applications/ShapeOptimizationApplication/custom_python/add_custom_utilities_to_python.cpp (hunk 1 above)
Comment (RezaNajian) [reply applications/ShapeOptimizationApplication/custom_python/add_custom_utilities_to_python.cpp]: corrected.

File: applications/ShapeOptimizationApplication/custom_utilities/response_functions/am_boundary_roughness_angle_response_function_utility.cpp [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+	KRATOS_CATCH("");
+}
+
+void AMBoundaryRoughnessAngleResponseFunctionUtility::CalculateGradient()
+{
+	KRATOS_TRY;
+	// First gradients are initialized
+	VariableUtils().SetHistoricalVariableToZero(SHAPE_SENSITIVITY, mrModelPart.Nodes());
+
+	for (auto& cond_i : mrModelPart.Conditions()){
```
Comment (philbucher) [inline applications/ShapeOptimizationApplication/custom_utilities/response_functions/am_boundary_roughness_angle_response_function_utility.cpp:68]: Check [here](https://github.com/KratosMultiphysics/Kratos/blob/90c385c74e8ab56ccacce6bd2eccd36c9f819575/applications/RomApplication/custom_strategies/custom_schemes/modal_derivative_scheme.hpp#L408-L458) if you want to do this in parallel as well

(basically you need to lock all the nodes of the Condition)

File: applications/ShapeOptimizationApplication/custom_utilities/response_functions/am_boundary_slope_angle_response_function_utility.cpp [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+#include "custom_utilities/response_functions/am_boundary_slope_angle_response_function_utility.h"
+#include "shape_optimization_application.h"
+#include "utilities/variable_utils.h"
+
+// ==============================================================================
+
+namespace Kratos
+{
+
+AMBoundarySlopeAngleResponseFunctionUtility::AMBoundarySlopeAngleResponseFunctionUtility(ModelPart& rModelPart, Parameters ResponseSettings)
```
Comment (philbucher) [inline applications/ShapeOptimizationApplication/custom_utilities/response_functions/am_boundary_slope_angle_response_function_utility.cpp:25]: browsing through this seems very similar to the other class

maybe you can refactor/unify?

File: applications/ShapeOptimizationApplication/python_scripts/analyzers/analyzer_internal.py [hunk 4]
```diff
@@ -122,6 +122,8 @@ def __CreateResponseFunctions( specified_responses, model ):
             "mesh_based_packaging",
             "surface_normal_shape_change",
             "face_angle",
+            "AM_boundary_slope_angle",
+            "AM_boundary_roughness_angle",
```
Comment (philbucher) [inline applications/ShapeOptimizationApplication/python_scripts/analyzers/analyzer_internal.py:126]:
```suggestion
            "am_boundary_slope_angle",
            "am_boundary_roughness_angle",
```

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/am_boundary_slope_angle.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
+# importing the Kratos Library
+import KratosMultiphysics as KM
+from KratosMultiphysics import Parameters, Logger
+from KratosMultiphysics.response_functions.response_function_interface import ResponseFunctionInterface
+import KratosMultiphysics.ShapeOptimizationApplication as KSO
+
+import time as timer
+
+
+def _AddConditionsFromParent(parent, child):
```
Comment (philbucher) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/am_boundary_slope_angle.py:10]: duplicated

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/response_function_factory.py [hunk 6]
```diff
@@ -25,6 +27,10 @@ def CreateResponseFunction(response_id, response_settings, model):
         return surface_normal_shape_change.SurfaceNormalShapeChange(response_id, response_settings, model)
     elif response_type == "face_angle":
         return face_angle.FaceAngleResponseFunction(response_id, response_settings, model)
+    elif response_type == "AM_boundary_slope_angle":
```
Comment (philbucher) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/response_function_factory.py:30]:
```suggestion
    elif response_type == "am_boundary_slope_angle":
```

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/response_function_factory.py [hunk 7]
```diff
@@ -25,6 +27,10 @@ def CreateResponseFunction(response_id, response_settings, model):
...
+        return am_boundary_slope_angle.AMBoundarySlopeAngleResponseFunction(response_id, response_settings, model)        
+    elif response_type == "AM_boundary_roughness_angle":
```
Comment (philbucher) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/response_function_factory.py:32]:
```suggestion
    elif response_type == "am_boundary_roughness_angle":
```

**Review by philbucher — COMMENTED** [review]
minor comment regarding code duplication otherwise nice work :+1: 

I leave the approval to @KratosMultiphysics/optimization as I am not familiar with the internals of the app

**Pull request closed without merging**